### PR TITLE
feat: support for AWS session token for S3

### PR DESCRIPTION
**Pull request by vigith**
Signed-off-by: Vigith Maurice <vigith@gmail.com>

I tested this using the following config

```yaml
apiVersion: dataflow.argoproj.io/v1alpha1
kind: Pipeline
metadata:
  annotations:
    dataflow.argoproj.io/owner: argoproj-labs
    dataflow.argoproj.io/test: 'true'
  name: session-test
spec:
  steps:
  - cat: {}
    name: main
    sinks:
    - log: {}
    sources:
    - s3:
        region: us-west-2       
        bucket: dataflow-s3-test
```

### CODE REVIEW

File: runner/sidecar/source/s3/s3.go [hunk 1]
```diff
@@ -45,10 +46,19 @@ func New(ctx context.Context, secretInterface corev1.SecretInterface, pipelineNa
 		}
 		secretAccessKey = string(secret.Data[x.Credentials.SecretAccessKey.Key])
 	}
+	var sessionToken string
+	{
+		secretName := x.Credentials.SessionToken.Name
+		secret, err := secretInterface.Get(ctx, secretName, metav1.GetOptions{})
+		// it is okay for sessionToken to be missing
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (vigith) [reply runner/sidecar/source/s3/s3.go:54]: added
```go
if !apierr.IsNotFound(err) {
   return nil, err
}
```

File: api/v1alpha1/aws_credentials.go [hunk 2]
```diff
@@ -5,4 +5,5 @@ import corev1 "k8s.io/api/core/v1"
 type AWSCredentials struct {
 	AccessKeyID     corev1.SecretKeySelector `json:"accessKeyId" protobuf:"bytes,1,opt,name=accessKeyId"`
 	SecretAccessKey corev1.SecretKeySelector `json:"secretAccessKey" protobuf:"bytes,2,opt,name=secretAccessKey"`
+	SessionToken    corev1.SecretKeySelector `json:"sessionToken" protobuf:"bytes,3,opt,name=sessionToken"`
```
Comment (whynowy) [inline api/v1alpha1/aws_credentials.go:8]: It's better to use pointers for `AccessKeyID`, `SecretAccessKey` and `SessionToken`.

File: runner/sidecar/source/s3/s3.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (whynowy) [reply runner/sidecar/source/s3/s3.go:54]: Using pointer to save 1 api call and avoid 404 error check.

**Review by alexec — APPROVED** [review]

File: runner/sidecar/sink/s3/s3.go [hunk 3]
```diff
@@ -44,10 +45,23 @@ func New(ctx context.Context, secretInterface v1.SecretInterface, x dfv1.S3Sink)
 		}
 		secretAccessKey = string(secret.Data[x.Credentials.SecretAccessKey.Key])
 	}
+	var sessionToken string
```
Comment (whynowy) [inline runner/sidecar/sink/s3/s3.go:48]: Same here, using pointer to simplify the logic.

Comment (vigith) [reply runner/sidecar/sink/s3/s3.go:48]: we need to revisit the AWS Credentials altogether. I kept it similar to the existing patterns.

**Pull request merged**